Mathlib/Data/Matrix/Notation.lean

```diff
@@ -114,6 +112,32 @@ macro_rules
     `(@Matrix.of (Fin $(quote semicolons.size)) (Fin 0) _ ![$emptyVecs,*])
   | `(!![$[,%$commas]*]) => `(@Matrix.of (Fin 0) (Fin $(quote commas.size)) _ ![])
 
+/-- Delaborator for the `!![]` notation. -/
+@[delab app.DFunLike.coe]
+def delabMatrixNotation : Delab := whenPPOption getPPNotation <| withOverApp 6 do
+  let em ← withAppFn <| withAppArg getExpr
+  let
+    .app
+      (.app
+        (.app (.const ``Matrix.of _) (.app (.const ``Fin _) em))
+        (.app (.const ``Fin _) en))
+      _:= em | failure
+  let some (m, n) ← withNatValue em fun m => withNatValue en fun n => pure (some (m, n)) | failure
+  if m = 0 then
+    withAppArg <| do
+      guard <| (← getExpr).isAppOf ``vecEmpty
+      let commas := mkArray n (Syntax.atom .none ",")
+      `(!![$[,%$commas]*])
+  else
+    if n = 0 then
+      withAppArg <| do
+        let `(![$[![]%$evecs],*]) ← delab | failure
```

Comment:
Matching on `delab` like this feels a big fragile, but I don't think it's actually wrong.

Reply:
I don't know if this design pattern is wrong because I don't have much experience writing delaborator.
I think it is better to ask @kmill.

Reply:
Matching on `delab` is at least done in the builtin delaborators.

Here's a refinement of the code, along with having it check that `pp.explicit` is false as well, since `DFunLike.coe` and `Matrix.of` have some implicit arguments.

```lean
@[delab app.DFunLike.coe]
def delabMatrixNotation : Delab := whenNotPPOption getPPExplicit <| whenPPOption getPPNotation <|
  withOverApp 6 do
    let mkApp3 (.const ``Matrix.of _) (.app (.const ``Fin _) em) (.app (.const ``Fin _) en) _ :=
      (← getExpr).appFn!.appArg! | failure
    let some m ← withNatValue em (pure ∘ some) | failure
    let some n ← withNatValue en (pure ∘ some) | failure
    withAppArg do
      if m = 0 then
        guard <| (← getExpr).isAppOfArity ``vecEmpty 1
        let commas := mkArray n (mkAtom ",")
        `(!![$[,%$commas]*])
      else
        if n = 0 then
          let `(![$[![]%$evecs],*]) ← delab | failure
          `(!![$[;%$evecs]*])
        else
          let `(![$[![$[$melems],*]],*]) ← delab | failure
          `(!![$[$[$melems],*];*])
```

Reply:
Thank you @kmill! I have rewritten.